library/core/src/iter/traits/iterator.rs

```diff
@@ -1825,10 +1825,20 @@ pub trait Iterator {
         Inspect::new(self, f)
     }
 
-    /// Borrows an iterator, rather than consuming it.
+    /// Iterator adapter that turns ownership-taking methods into mutating methods.
     ///
-    /// This is useful to allow applying iterator adapters while still
-    /// retaining ownership of the original iterator.
+    /// Most `Iterator` methods, other than the `next` method (and this one),
+    /// take ownership (via a `self` parameter) of the iterator and then mutate it.
+    /// This method retuns a mutable reference to the original iterator,
+    /// that acts like the original iterator in all respects,
+    /// except that ownership-taking methods take ownership of the reference instead of of the original.
+    /// All mutation by ownership-taking methods will still mutate the original iterator.
+    ///
+    /// # Technical details
+    ///
+    /// The mutable reference that this method returns, implements the `Iterator` trait via:
+    /// [impl<I: Iterator + ?Sized> Iterator for &mut I { type Item = I::Item; ...}](trait.Iterator.html#impl-Iterator-for-%26mut I).
+    /// This implementation simply passes all method calls on to the original iterator.
```

Comment:
```suggestion
    /// Most `Iterator` adapters take ownership (via a `self` parameter) of the iterator.
    /// This method returns a mutable reference to the original, which implements the
    /// iterator trait by delegating to the original iterator. This allows iterator adapters
    /// to take ownership of the reference, mutating the original iterator while retaining
    /// ownership.
```

- It's not true that only `next` doesn't take ownership, there are other methods like `nth` and `advance_by`.
- "and mutate it" isn't exactly true, because the methods themselves don't mutate the iterator. We can just omit this.
- "mutation by ownership-taking methods" is a little awkward, I updated the wording to use "adapters" to refer to the adapter methods explicitly.
- The hyperlink to the iterator implementation doesn't seem to work: https://github.com/rust-lang/rust/issues/105600. I added a written explanation instead.

Does that seem reasonable?

Reply:
> * It's not true that only `next` doesn't take ownership, there are other methods like `nth` and `advance_by`.

That's why the text says "Most .. methods", but I thought it important to mention `next` as an exception because it is so fundamental. Of the [76 methods listed](https://doc.rust-lang.org/nightly/std/iter/trait.Iterator.html) only 17 don't take `self`, so 59 methods do, which certainly counts as most, no? But I would also be happy with "many" instead of "most"... I see that your proposed replacement differs mostly by replacing "methods" with "adapters" but all of, and `advance_by` are adapters and `next` and `nth` arguably even "iterator adapters" (because Option impls Iterator). 

> * "and mutate it" isn't exactly true, because the methods themselves don't mutate the iterator. We can just omit this.

It is true! They take ownership and then you have no way of knowing what they do, but in fact they do mutate the iterator, which you can see if you use by_ref to prevent ownership-taking! I think it is important to mention this, because then it cannot be a surprise if after preventing ownership taking these methods still (!!!) mutate the original. They always mutate the original. The only difference is in whether ownership is transferred or not.

> * "mutation by ownership-taking methods" is a little awkward, I updated the wording to use "adapters" to refer to the adapter methods explicitly, which I explained take ownership.

The book calls `next` an adapter and it does not take ownership. The book names consuming adapters (next and other methods that call next), iterator adapters (methods whose return value impls iterator (next returns an Option which is an iterator, so arguably it is also an iterator adapter)), and other methods (only size_hint?).

> * The hyperlink to the iterator implementation doesn't seem to work:

Should work now... thanks for catching!

Reply:
> That's why the text says "Most .. methods", but I thought it important to mention next as an exception because it is so fundamental.

My only concern was that "other than the `next` method (and this one)" seems to suggest that every other iterator method is consuming.

> It is true! They take ownership and then you have no way of knowing what they do, but in fact they do mutate the iterator, which you can see if you use by_ref to prevent ownership-taking

Right, but the methods themselves don't mutate it, they simply return a new iterator adapter/wrapper.

> The book calls next an adapter and it does not take ownership. The book names consuming adapters (next and other methods that call next), iterator adapters (methods whose return value impls iterator (next returns an Option which is an iterator, so arguably it is also an iterator adapter)), and other methods (only size_hint?).

Sure, but we already clarified that we are talking about adapters that consume `self` in the previous sentence.